src/ast/passes/semantic_analyser.cpp

```diff
@@ -3598,11 +3598,28 @@ void SemanticAnalyser::visit(AttachPoint &ap)
       if (ap.mode[i - 1] == ap.mode[i])
         ap.addError() << "watchpoint modes may not be duplicated";
     }
-    const auto invalid_modes = arch::invalid_watchpoint_modes();
-    if (std::ranges::any_of(invalid_modes,
-
-                            [&](const auto &mode) { return mode == ap.mode; }))
-      ap.addError() << "invalid watchpoint mode: " << ap.mode;
+    const auto &modes = arch::watchpoint_modes();
+    if (!modes.contains(ap.mode)) {
+      if (modes.empty()) {
+        // There are no valid modes.
+        ap.addError() << "watchpoints not supported";
+      } else {
+        // Build a suitable error with hint.
+        auto &err = ap.addError();
+        err << "invalid watchpoint mode: " << ap.mode;
+        auto &hint = err.addHint();
+        hint << "supported modes: ";
+        bool first = true;
+        for (const auto &mode : modes) {
+          if (!first) {
+            hint << ", " << mode;
+          } else {
+            hint << mode;
+            first = false;
+          }
+        }
```

Comment:
We have `utils::str_join`, can we use it?

Reply:
Using it now. Although it hurts my soul a bit to see an `O(n^2)` string join :) We should fix that separately